### PR TITLE
Support FFTs

### PR DESCRIPTION
**Pull request by mcabbott**
Since @oxinabox was complaining of having nothing to review... this may need a little updating and a little polish. 

The idea was that FFT ought to be written like `fft(A, :time => :freq)` since obviously those aren't the same thing, and you would never want to end taking a dot product of them etc. And if you don't specify a name, then `:x∿` is a character that behaves well, I think.

### CODE REVIEW

File: src/functions_math.jl
Comment (oxinabox) [inline src/functions_math.jl]: Thoughs on the pro and con of using kwargs to rename?
```
fft((A;  x=:xf, y=:yf)
```

File: src/functions_math.jl
Comment (oxinabox) [inline src/functions_math.jl]: Per BlueStyle https://github.com/invenia/BlueStyle/blob/master/.mailmap#method-definitions

> Only use short-form function definitions when they fit on a single line

I suspect this code was written before we adopted the style-guide ?

File: src/functions_math.jl
Comment (oxinabox) [inline src/functions_math.jl]: this should move to `src/compat.jl`
Or possibly we can just make use of the **Compat.jl** package for it?

File: src/name_core.jl
Comment (oxinabox) [inline src/name_core.jl]: Can we move these into the tests?

**Comment by oxinabox** [comment]
This would be the first time this package actually decided on a name for a thing on its own.
Its never done that in any other PR before.
Only ever inserted the wildcard name.

I think the naming done in this PR make sense though.

My only worry is if we could be creating name conflicts (admittedly the names would have had to be pretty weird to end up in that situation).

---
unrelated:

It seems like this name actually works fine as a kwarg which is good.
```
julia> foo(;kwargs...) = kwargs
foo (generic function with 1 method)

julia> foo(x∿=4)
pairs(::NamedTuple) with 1 entry:
  :x∿ => 4
```

File: src/functions_math.jl
Comment (mcabbott) [reply src/functions_math.jl]: I can just always use ` Base.sym_in` inline. (I moved it out to see if it mattered for speed.)

The treatment of `plan_fft` seems a bit awkward to me. The alternative it to return a new struct `NamedPlan` or something, which would explicitly encode the names & dimensions, and make its application zero-cost again. But perhaps this adds more code than it deserves, thoughts?

File: src/functions_math.jl
Comment (mcabbott) [reply src/functions_math.jl]: Or was doodled up for my own amusement. Easy to fix.

File: src/name_core.jl
Comment (mcabbott) [reply src/name_core.jl]: Sure

File: src/functions_math.jl
Comment (mcabbott) [reply src/functions_math.jl]: I thought about that, but concluded that it was too confusing about which one is the old & the new name.

File: src/functions_math.jl
Comment (oxinabox) [reply src/functions_math.jl]: > I can just always use Base.sym_in inline. (I moved it out to see if it mattered for speed.)

That would be worse for readability.
just a 
```
if isdefined(Base, :hasproperty)
    hasproperty(x, s)::Bool = Base.sym_in(s, propertynames(x))
end
```

in the compat file.
(or check Compat.jl)

**Comment by mcabbott** [comment]
Yes I was quite pleased to find `∿`, it's relevant, you can type it, and it's an ordinary letter as far as Julia's concerned. I can't find a single example of anyone using it in code, so I doubt there will be collisions.

File: src/functions_math.jl
Comment (oxinabox) [reply src/functions_math.jl]: I think `NamedPlan` make sense, but i don't know FFT well

Comment (mcabbott) [reply src/functions_math.jl]: OK will tidy up the struct version then.

Comment (mcabbott) [reply src/functions_math.jl]: Will get to this eventually, BTW. The struct version had a few unexpected complications.

**Comment by mcabbott** [comment]
BTW, the status here is that the version with `struct NamedPlan` kept getting more complicated, and I wasn't exactly sure what its rules should be. (I never pushed this, it seems.)

The simpler version here with `Base.:*(plan::AbstractFFTs.$plan_type, A::NamedDimsArray{L,T,N})` (in which the plan can't explicitly carry names) can probably be improved using  https://github.com/JuliaMath/FFTW.jl/pull/119, which makes the dims on which it acts part of the type.

File: src/functions_math.jl
Comment (mcabbott) [reply src/functions_math.jl:231]: OK, I trashed the whole `NamedPlan` business, it got complicated. We can re-invent it if someone uses this & needs it, maybe make the basic case work first.

**Comment by mcabbott** [comment]
Maybe this is done. 

It has only iffy support for `plan_fft`, and none at all for `mul!(out, plan, input)` or `fft!` or `plan_fft!`. I guess `fft!` would not be hard to add, but the others... can be done later if someone wants them.

File: src/name_core.jl
Comment (mcabbott) [reply src/name_core.jl]: This is now deleted, in favour of `_rename` from another PR.

File: src/functions_math.jl
Comment (oxinabox) [inline src/functions_math.jl:143]: Should this be  called `extra_args` or something?

File: src/functions_math.jl
Comment (oxinabox) [inline src/functions_math.jl:253]: This commented out section probable needs to be deleted?

File: src/functions_math.jl
Comment (oxinabox) [inline src/functions_math.jl:137]: Maybe we should move this all to its own file?
and then the test to match?

File: src/functions_math.jl
Comment (oxinabox) [inline src/functions_math.jl:142]:
```suggestion
    if fun in (:irfft, :brfft)  # These take one more argument, a size
```

Two spaces before end of line comments, per BlueStyle

**Review by oxinabox — APPROVED** [review]
Ok I am once again getting toward the bottom of myh to review stack (though i think there are still a dozen things below this)

This LGTM.
I have granted you merge rights on the repo.
so you can merge this when ready.
All the usual [ColPrac](https://github.com/SciML/ColPrac) things apply about not merging your own PR without review (give ColPrac a read if you haven't already)

File: src/functions_math.jl
Comment (mcabbott) [reply src/functions_math.jl:253]: Will do.

File: src/functions_math.jl
Comment (mcabbott) [reply src/functions_math.jl]: I changed it, to also include a string to insert into docstrings.

File: src/functions_math.jl
Comment (mcabbott) [reply src/functions_math.jl]: Done.

**Comment by mcabbott** [comment]
OK, thanks! 

If I understand that right I'm supposed to merge this, although perhaps I wait until solid CI. Will JuliaRegistrator also take orders from me now?

**Comment by nickrobinson251** [comment]
> I'm supposed to merge this

Yeah, merge your own PRs once they're approved and CI passes, then register the new version (fingers crossed Registrator will work for you now) :) 

thanks!

**Comment by oxinabox** [comment]
Urg, i bet Registrator won't work.

**Comment by mcabbott** [comment]
Done! 

#149 has a version bump & is approved, I will merge it too & then this can be tagged.

**Pull request merged**
